### PR TITLE
Update Node environment detection

### PR DESCRIPTION
**Pull request by rajsite**
### Description

Updates the heuristic used by pyodide to detect if it is running in a node context.
The benefits of this change:

1. Better alignment with the heuristic already generated by [emscripten](https://github.com/emscripten-core/emscripten/blob/6f0238367348f5e33605b3f0a42455fb3a2d739f/src/shell.js#L116).
2. Unblocks usage of pyodide in Deno via the [`npm: specifiers`](https://deno.com/manual@v1.33.1/node/npm_specifiers) compatibility feature. See discussion in https://github.com/pyodide/pyodide/issues/3420#issuecomment-1524741406.

### Checklists

- Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
  - Removed checkbox reason: This change on it's own seems potentially too minor for the Changelog. I can absolutely document this heuristic change in the Changelog if desired. Seems like a more interesting entry is experimental Deno support when smoke test and docs are merged.
- Add / update tests
  - Removed checkbox reason: Relying on existing tests for current behavior. Tests for Deno support via `npm:` registry specifiers would be a follow-up
- Add new / update outdated documentation
  - Removed checkbox reason: Similar to above, docs are a follow-up

### CODE REVIEW

**Review by hoodmane — APPROVED** [review]

File: src/js/compat.ts [hunk 1]
```diff
@@ -2,9 +2,9 @@
 declare var process: any;
 
 export const IN_NODE =
-  typeof process === 'object' &&
-  typeof process.versions === 'object' &&
-  typeof process.versions.node === 'string' &&
+  typeof process === "object" &&
```
Comment (rajsite) [inline src/js/compat.ts:5]: good bot 🤖🎉

**Comment by rajsite** [comment]
@hoodmane if it's not too much trouble, after merging could this change be included in a prerelease package to actually verify the end-to-end workflow with Deno?

**Comment by hoodmane** [comment]
I'm going to release a version 0.23.2 so if neither @rth nor @ryanking13 veto it, I could backport it into that.

**Comment by rth** [comment]
Sounds good to me as well. Thanks @rajsite !

**Pull request merged**